lib/Sema/TypeCheckDecl.cpp

```diff
@@ -3075,14 +3075,40 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
           .fixItReplace(ED->getExtendedTypeRepr()->getSourceRange(),
                         canExtType->getString());
       } else if (!wasAlreadyInvalid) {
-		// If nothing else applies, fall back to a generic diagnostic.
+        // If nothing else applies, fall back to a generic diagnostic.
         ED->diagnose(diag::non_nominal_extension, extType);
       }
       return;
     }
 
     TC.validateExtension(ED);
 
+    extType = ED->getExtendedType();
+    if (extType && !extType->hasError()) {
+      // The first condition catches syntactic forms like
+      //     protocol A & B { ... } // may be protocols or typealiases
+      // The second condition also looks through typealiases and catches
+      //    typealias T = P1 & P2 // P2 is a refined protocol of P1
+      //    extension T { ... }
+      // However, it is trickier to catch cases like
+      //    typealias T = P2 & P1 // P2 is a refined protocol of P1
+      //    extension T { ... }
+      // so we don't do that here.
+      auto extTypeRepr = ED->getExtendedTypeRepr();
+      auto *extTypeNominal = extType->getAnyNominal();
+      bool firstNominalIsNotMostSpecific =
+        extTypeNominal && extTypeNominal != nominal;
+      if (isa<CompositionTypeRepr>(extTypeRepr)
+          || firstNominalIsNotMostSpecific) {
+        auto replacementType = nominal->getDeclaredType();
+        ED->diagnose(diag::composition_in_extended_type, replacementType)
+          .highlight(extTypeRepr->getSourceRange());
+        if (firstNominalIsNotMostSpecific)
+          ED->diagnose(diag::composition_in_extended_type_alternative,
+                       extTypeNominal->getDeclaredType());
+      }
```

Comment:
Fix-its? On both branches of the `if`, I guess.

Reply:
Shouldn't we _not_ be providing a fix-it since there isn't one clear right answer? :confused:

Reply:
A fix-it on a note counts as an option, and it's definitely the most *likely* right answer in the case where you print the note. When you don't print the note (because the compiler picked the most specific type anyway), it's almost certainly the right answer.